utils/src/bloom_filters/rolling_bloom_filter/mod.rs

```diff
@@ -0,0 +1,116 @@
+/// with the specified number of items and expected false positive rate.
+///
+/// It works by constructing 3 smaller bloom filters and inserting items until all subfilters are full.
+/// Then the oldest subfilter is replaced by a new one (with new seed values to prevent persistent false positives).
+/// An element is reported as present if it's present in any subfilter.
+///
+/// Used memory per element (only hashes are stored, so it does not depend on the size of the input elements):
+/// Memory (bytes)  FPP
+/// 3.123           0.001
+/// 4.023           0.0001
+/// 4.920           0.00001
+/// 5.820           0.000001
+
+pub struct RollingBloomFilter<T> {
+    /// The list of subfilters that store items (newest first, oldest last)
+    subfilters: CyclicFilter<BloomFilter<T>, SUBFILTER_COUNT>,
+
+    /// Number of items added to the current subfilter
+    subfilter_inserted_count: usize,
+
+    /// Maximum number of items that can be added to a subfilter
+    subfilter_inserted_max: usize,
+
+    /// Desired false positive probability in each subfilter
+    fpp_subfilter: f64,
+}
+
+impl<T: Hash> RollingBloomFilter<T> {
+    /// Constructs a new rolling bloom filter.
+    ///
+    /// # Arguments
+    /// `size` - maximum number of items to store in the filter (up to 3/2*size of last items can be stored at times)
+    /// `fpp` - required false positive rate
+    /// `rng` - random number generator used to randomize hashes
+    pub fn new(size: usize, fpp: f64, rng: &mut impl Rng) -> Self {
+        assert!(size > 0);
+        assert!(fpp > 0.0 && fpp < 1.0);
+
+        // Use the smaller bloom filters to store size/2 of the last items in each.
+        // This way we always remember `size..3/2*size` items.
+        let subfilter_inserted_max = std::cmp::max(1, size / (SUBFILTER_COUNT - 1));
+
+        // `fpp_per_filter` must be derived from `fpp` so that the probability of being detected in any subfilter is about `fpp`.
+        // Since each subfilter can be considered independent, it is about (1 - p)^3,
+        // which can be approximated by 1-3*p (if p is small).
+        // As a result, the required ffp per filter should be about fpp / 3.0.
+        let fpp_subfilter = fpp / SUBFILTER_COUNT as f64;
+
+        let subfilters = CyclicFilter::<BloomFilter<T>, SUBFILTER_COUNT>::new(BloomFilter::new(
+            subfilter_inserted_max,
+            fpp_subfilter,
+            rng,
+        ));
+
+        RollingBloomFilter {
+            subfilters,
+            subfilter_inserted_count: 0,
+            subfilter_inserted_max,
+            fpp_subfilter,
+        }
+    }
+
+    /// Add item to the rolling bloom filter.
+    pub fn insert(&mut self, item: &T, rng: &mut impl Rng) {
+        debug_assert!(self.subfilter_inserted_count < self.subfilter_inserted_max);
+        // Insert the element into the newest subfilter
+        self.subfilters.get_current_mut().insert(item);
```

Comment:
If all you need to do is insert, then `get_current_mut` isn't necessary. Just create a function `insert` in `CyclicFilter`

Reply:
This is not possible because `CyclicFilter` is now more like a generic container and doesn't know anything about bloom filters.